clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3442,147 +3488,132 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
-                           "CollectPossibleKernelAttributes");
-        }
+  case attr::Kind::ReqdWorkGroupSize: {
+    auto *RWGSA = cast<ReqdWorkGroupSizeAttr>(A);
+    if (auto *Existing = SYCLKernel->getAttr<ReqdWorkGroupSizeAttr>()) {
+      ASTContext &Ctx = S.getASTContext();
+      if (Existing->getXDimVal(Ctx) != RWGSA->getXDimVal(Ctx) ||
+          Existing->getYDimVal(Ctx) != RWGSA->getYDimVal(Ctx) ||
+          Existing->getZDimVal(Ctx) != RWGSA->getZDimVal(Ctx)) {
+        S.Diag(SYCLKernel->getLocation(),
+               diag::err_conflicting_sycl_kernel_attributes);
+        S.Diag(Existing->getLocation(), diag::note_conflicting_attribute);
+        S.Diag(RWGSA->getLocation(), diag::note_conflicting_attribute);
+        SYCLKernel->setInvalidDecl();
+      }
+    } else if (auto *Existing =
+                   SYCLKernel->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+      ASTContext &Ctx = S.getASTContext();
+      if (Existing->getXDimVal(Ctx) < RWGSA->getXDimVal(Ctx) ||
+          Existing->getYDimVal(Ctx) < RWGSA->getYDimVal(Ctx) ||
+          Existing->getZDimVal(Ctx) < RWGSA->getZDimVal(Ctx)) {
+        S.Diag(SYCLKernel->getLocation(),
+               diag::err_conflicting_sycl_kernel_attributes);
+        S.Diag(Existing->getLocation(), diag::note_conflicting_attribute);
+        S.Diag(RWGSA->getLocation(), diag::note_conflicting_attribute);
+        SYCLKernel->setInvalidDecl();
+      } else {
+        SYCLKernel->addAttr(A);
       }
+    } else {
+      SYCLKernel->addAttr(A);
+    }
+    break;
+  }
+  case attr::Kind::SYCLIntelMaxWorkGroupSize: {
+    auto *SIMWGSA = cast<SYCLIntelMaxWorkGroupSizeAttr>(A);
+    if (auto *Existing = SYCLKernel->getAttr<ReqdWorkGroupSizeAttr>()) {
+      ASTContext &Ctx = S.getASTContext();
+      if (Existing->getXDimVal(Ctx) > SIMWGSA->getXDimVal(Ctx) ||
+          Existing->getYDimVal(Ctx) > SIMWGSA->getYDimVal(Ctx) ||
+          Existing->getZDimVal(Ctx) > SIMWGSA->getZDimVal(Ctx)) {
+        S.Diag(SYCLKernel->getLocation(),
+               diag::err_conflicting_sycl_kernel_attributes);
+        S.Diag(Existing->getLocation(), diag::note_conflicting_attribute);
+        S.Diag(SIMWGSA->getLocation(), diag::note_conflicting_attribute);
+        SYCLKernel->setInvalidDecl();
+      } else {
+        SYCLKernel->addAttr(A);
+      }
+    } else {
+      SYCLKernel->addAttr(A);
     }
+    break;
+  }
+  case attr::Kind::SYCLIntelKernelArgsRestrict:
+  case attr::Kind::SYCLIntelNumSimdWorkItems:
+  case attr::Kind::SYCLIntelSchedulerTargetFmaxMhz:
+  case attr::Kind::SYCLIntelMaxGlobalWorkDim:
+  case attr::Kind::SYCLIntelNoGlobalWorkOffset:
+  case attr::Kind::SYCLIntelUseStallEnableClusters:
+  case attr::Kind::SYCLIntelLoopFuse:
+  case attr::Kind::SYCLIntelFPGAMaxConcurrency:
+  case attr::Kind::SYCLIntelFPGADisableLoopPipelining:
+  case attr::Kind::SYCLIntelFPGAInitiationInterval:
+  case attr::Kind::SYCLSimd: {
+    if ((A->getKind() == attr::Kind::SYCLSimd) && KernelBody &&
```

Comment:
This is a bit messy -- put this logic in the `attr::Kind::SYCLSimd` case, then use `LLVM_FALLTHROUGH`?

Reply:
Oof... i didn't notice that when moving this.  THAT is simple enough and offensive enough I'll fix it :)